src/calliope/config.py

```diff
@@ -0,0 +1,369 @@
+
+        Args:
+            update_dict (dict): Dictionary with which to update the base model.
+            deep (bool, optional): Set to True to make a deep copy of the model. Defaults to False.
+
+        Returns:
+            BaseModel: New model instance.
+        """
+        new_dict: dict = {}
+        # Iterate through dict to be updated and convert any sub-dicts into their respective pydantic model objects
+        for key, val in update_dict.items():
+            key_class = getattr(self, key)
+            if isinstance(key_class, ConfigBaseModel):
+                new_dict[key] = key_class.update(val)
+                key_class._kwargs = val
+            else:
+                new_dict[key] = val
+        updated = super().model_copy(update=new_dict, deep=deep)
+        updated.model_validate(updated)
+        self._kwargs = update_dict
+        return updated
+
+    @overload
+    def model_yaml_schema(self, filepath: str | Path) -> None: ...
+
+    @overload
+    def model_yaml_schema(self, filepath: None = None) -> str: ...
+
+    def model_yaml_schema(self, filepath: str | Path | None = None) -> None | str:
+        """Generate a YAML schema for the class.
+
+        Args:
+            filepath (str | Path | None, optional): If given, save schema to given path. Defaults to None.
+
+        Returns:
+            None | str: If `filepath` is given, returns None. Otherwise, returns the YAML string.
+        """
+        # By default, the schema uses $ref/$def cross-referencing for each pydantic model class,
+        # but this isn't very readable when rendered in our documentation.
+        # So, we resolve references and then delete all the `$defs`
+        schema_dict = AttrDict(jsonref.replace_refs(self.model_json_schema()))
+        schema_dict.del_key("$defs")
+        return schema_dict.to_yaml(filepath)
+
+    @property
+    def applied_keyword_overrides(self) -> dict:
+        """Most recently applied keyword overrides used to update this configuration.
+
+        Returns:
+            dict: Description of applied overrides.
+        """
+        return self._kwargs
+
+
+class ModeBaseModel(ConfigBaseModel):
+    """Mode-specific configuration, which will be hidden from the string representation of the model if that mode is not activated."""
+
+    mode: MODES_T = Field(default="plan")
+    """Mode in which to run the optimisation."""
+
+    @model_validator(mode="after")
+    def update_repr(self) -> Self:
+        """Hide config from model string representation if mode is not activated."""
+        for key, val in self.model_fields.items():
+            if key in get_args(MODES_T):
+                val.repr = self.mode == key
+        return self
+
+
+class Init(ConfigBaseModel):
+    """All configuration options used when initialising a Calliope model."""
+
+    model_config = {
+        "title": "init",
+        "extra": "forbid",
+        "frozen": True,
+        "json_schema_extra": hide_from_schema(["def_path"]),
```

Comment:
Because the schema would only be created on-the-fly to create relevant documentation pages  (easier to do with jsonschema2md than from pydantic directly). We don't want `def_path` in that because it isn't an option a user can actually supply directly.

Reply:
Understood. As a counterpoint, users will still see it when they look at this in the code, no?
Would specifying in the docs that this is auto-set by us be sufficient? It would reduce code complexity, and perhaps even reduce user confusion.